// jobportal-backend/src/controllers/applicationController.js
import { asyncHandler } from '../utils/asyncHandler.js'
import { sendCreated, sendSuccess } from '../utils/response.js'
import { createApplication, deleteApplicationById, getApplicationById, listApplications, updateApplicationById } from '../services/applicationService.js'
import { getUserById, incrementUserStats } from '../services/userService.js'
import { Job } from '../models/Job.js'
import { AppError } from '../utils/AppError.js'

/**
 * @route POST /api/applications
 * @description Un candidato aplica a una vacante.
 */
export const createApplicationController = asyncHandler(async (req, res) => {
  const candidateId = req.body.candidate || req.user?.id
  if (!candidateId) {
    throw new AppError('Usuario no identificado', 400)
  }

  const jobDoc = await Job.findById(req.body.job).select('company').lean()
  if (!jobDoc) {
    throw new AppError('Vacante no encontrada', 404)
  }

  if (req.user?.role === 'employer') {
    const currentUser = await getUserById(req.user.id)
    const employerCompanyId = currentUser?.company?._id?.toString() || currentUser?.company?.id
    if (employerCompanyId && jobDoc.company?.toString() === employerCompanyId) {
      throw new AppError('No puedes postularte a tus propias vacantes', 400)
    }
  }

  const payload = {
    job: req.body.job,
    candidate: candidateId
  }

  if (req.body.message) {
    payload.message = req.body.message
  }

  if (req.file) {
    const publicBase = `${req.protocol}://${req.get('host')}`
    payload.cv = {
      filename: req.file.filename,
      originalName: req.file.originalname,
      mimeType: req.file.mimetype,
      url: `${publicBase}/uploads/cv/${req.file.filename}`
    }
  }

  const application = await createApplication(payload)
  const updatedUser = await incrementUserStats(candidateId, { applicationsSent: 1 })
  return sendCreated(res, {
    message: 'Postulación enviada',
    data: {
      application,
      stats: updatedUser?.stats || null
    }
  })
})

/**
 * @route GET /api/applications
 * @description Lista postulaciones filtradas por usuario o vacante.
 */
export const listApplicationsController = asyncHandler(async (req, res) => {
  const applications = await listApplications(req.query)
  return sendSuccess(res, { data: applications })
})

/**
 * @route GET /api/applications/:id
 * @description Detalle completo de una postulación.
 */
export const getApplicationController = asyncHandler(async (req, res) => {
  const application = await getApplicationById(req.params.id)
  return sendSuccess(res, { data: application })
})

/**
 * @route PATCH /api/applications/:id
 * @description Actualiza estado o mensaje de la postulación.
 */
export const updateApplicationController = asyncHandler(async (req, res) => {
  const application = await updateApplicationById(req.params.id, req.body)
  return sendSuccess(res, { message: 'Postulación actualizada', data: application })
})

/**
 * @route DELETE /api/applications/:id
 * @description Elimina una postulación.
 */
export const deleteApplicationController = asyncHandler(async (req, res) => {
  await deleteApplicationById(req.params.id)
  return sendSuccess(res, { message: 'Postulación eliminada' })
})
